### ISSUE TITLE
Missing "Maven > Update Maven Project" context menu on Maven projects

### ISSUE THREAD

**Issue by mickaelistria** [eclipse-m2e/m2e-core#724]
Using latest snapshots, Missing "Maven > Update Maven Project" context menu on Maven projects

**Comment by laeubi** [comment]
I think the context menu is named just named `Update Project...` but I can't even find that name anywhere. Maybe this was part of a bundle that was removed?

**Comment by mickaelistria** [comment]
The menu is present with latest m2e release (org.eclipse.m2e.core.feature 1.20.1.20220227-1319), here is what menu spy shows
![Screenshot from 2022-05-23 17-24-05](https://user-images.githubusercontent.com/469058/169853820-40663562-3ae8-42b7-a671-7d8d4d2483d5.png)

**Comment by laeubi** [comment]
Do you mean "Update Maven Project" in the Menu bar (Project > Update Maven Project) or in the Context Menu wen you right click a pom file (Maven > Update Project...)?
The first is (always?) shown, the second is missing in current snapshot as far as I can see.

**Comment by mickaelistria** [comment]
Context-menu.

**Comment by mickaelistria** [comment]
Other part of the removal is at https://github.com/eclipse-m2e/m2e-core/commit/f7a928179f93874ec7659122e54044e841125f34 . Athough this method is marked as Deprecated, it's actually still useful and internal, so let's restore it too.

**Comment by HannesWell** [comment]
Looks like I oversaw that.

@mickaelistria do you plan to work on this? Or should I resort that part internally?

**Comment by mickaelistria** [comment]
I'm stopping my laptop for today. If you want to provide a patch before I'm back to work tomorrow morning, it's very welcome ;) Otherwise I'll take care of it tomorrow.

**Comment by HannesWell** [comment]
> I'm stopping my laptop for today. If you want to provide a patch before I'm back to work tomorrow morning, it's very welcome ;) Otherwise I'll take care of it tomorrow.

I created https://github.com/eclipse-m2e/m2e-core/pull/733 to basically revert the removal. I also tried to replace the deprecated extensions of the a `org.eclipse.ui.menus` extension-point using a `menuContribution` but failed for today because I don't know these extensions well. I think the most difficult part is to find a proper `locationURI`. I tried `locationURI="popup:org.eclipse.ui.popupMenus"` but did not work.

Anyway. I pushed the branch to the m2e-core repo directly, so in case you want, you can improve that PR further (from the last change in that area I believe you are much more proficient with that than I am).

**Comment by laeubi** [comment]
> I think the most difficult part is to find a proper `locationURI`

https://www.vogella.com/tutorials/EclipseCodeAccess/article.html#menu-spy

**Comment by mickaelistria** [comment]
> I think the most difficult part is to find a proper locationURI

I spend a non-negligible amount of time on this exact same issue, in vain. So I fully sympathize with that and am happy with keeping the objectContribution so far.

**Comment by laeubi** [comment]
> I spend a non-negligible amount of time on this exact same issue, in vain.

Probably either MenuSpy, Eclipse Debugging or E4Tooling or even all of them must then be enhanced to better support this? I also found it frustrating in the past because there is little to no feedback and simply *nothing* happens...

**Comment by mickaelistria** [comment]
> Probably either MenuSpy, Eclipse Debugging or E4Tooling or even all of them must then be enhanced to better support this?

Yes, probably. I can get MenySpy working and retrieve a locationUri, but then trying to use this locationUri in a menu does nothing interesting. I'm not sure it's a tooling issue, it's possible that menus creates with the outdated "popupMenus" extension points are actually not extensible by commands.

**Comment by HannesWell** [comment]
> > I think the most difficult part is to find a proper locationURI
> 
> I spend a non-negligible amount of time on this exact same issue, in vain. So I fully sympathize with that and am happy with keeping the objectContribution so far.


There are many more deprecated extension points used, so if we know how to replace some, we can then replace all and fix the deprecation warnings.

> > I spend a non-negligible amount of time on this exact same issue, in vain.
> 
> Probably either MenuSpy, Eclipse Debugging or E4Tooling or even all of them must then be enhanced to better support this? I also found it frustrating in the past because there is little to no feedback and simply _nothing_ happens...

Indeed. Better debugging/error reporting support would be very helpful.

**Comment by mickaelistria** [comment]
>  if we know how to replace some, we can then replace all and fix the deprecation warnings.

The context-menu Project Explorer should be extensible with "org.eclipse.ui.menus" and "org.eclipse.ui.navigator.viewerContent" extension-points I believe. But first step would be to turn all actions into Commands/Handlers first and make the "legacy" actions used in objectContributions delegate to the command; then we can remove the objectContributions and use a more modern extension point. But it's more work for no obvious functional added-value (only functional gain would be commands appearing in Ctrl+3)...
